File: Example/HighFit/HighFit/HighFit/DashboardViewController.m
//
//  DashsboardViewController.m
//  HighFit
//
//  License: www.highcharts.com/license
//  Copyright © 2024 Highsoft AS. All rights reserved.
//

#import "DashboardViewController.h"
#import <Highcharts/Highcharts.h>
#import "OptionsProvider.h"
#import "DataTableViewController.h"

@interface DashboardViewController ()
@property (strong, nonatomic) NSMutableArray *sources;
@property (strong, nonatomic) NSArray *data;
@property (strong, nonatomic) NSString *dataName;
@property (strong, nonatomic) NSMutableArray *charts;
@end

@implementation DashboardViewController

+ (instancetype)sharedDashboard
{
    static DashboardViewController *sharedDashboard = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        sharedDashboard = [[DashboardViewController alloc] initWithNibName:nil bundle:nil];
    });
    
    return sharedDashboard;
}

- (instancetype)initWithNibName:(NSString *)nibNameOrNil bundle:(NSBundle *)nibBundleOrNil
{
    self = [super initWithNibName:nibNameOrNil bundle:nibBundleOrNil];
    if (self) {
        
        self.title = @"Dashboard";
        
        self.tabBarItem.image = [UIImage imageNamed:@"ic_content_paste_white"];
        
        [self loadSourcesAndData];
        
        self.tableView.rowHeight = 240.0f;
        
        self.tableView.tableHeaderView = [[UIView alloc] initWithFrame:CGRectMake(0.0f, 0.0f, self.tableView.bounds.size.width, 0.01f)];

        self.dataName = @"day";
        
        [self.segment addTarget:self action:@selector(actionSegment:) forControlEvents:UIControlEventValueChanged];
        
        self.tableView.tableHeaderView = nil;
    }
    return self;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view from its nib.
    
    self.tableView.tableFooterView = [[UIView alloc] initWithFrame:CGRectMake(0, 0, self.view.frame.size.width, 8.0f)];
    
    
    self.tableView.separatorColor = [UIColor clearColor];
    self.tableView.separatorInset = UIEdgeInsetsZero;

}

- (void)viewDidAppear:(BOOL)animated {
    [super viewDidAppear:animated];
    
    [self.tableView reloadData];
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

/*
#pragma mark - Navigation

// In a storyboard-based application, you will often want to do a little preparation before navigation
- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    // Get the new view controller using [segue destinationViewController].
    // Pass the selected object to the new view controller.
}
*/

#pragma mark - Table view data source

- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath
{
    return 250.0f;
}

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return self.data.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    
    NSMutableDictionary *options = [NSMutableDictionary dictionaryWithDictionary:[self.sources objectAtIndex:indexPath.row]];
    
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:options[@"chartType"]];
    
    // Configure the cell...
    if (cell == nil) {
        
        cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:options[@"chartType"]];
    
        cell.selectionStyle = UITableViewCellSelectionStyleNone;
    
        HIChartView *chartView = [[HIChartView alloc] initWithFrame:CGRectMake(5.0f, 5.0f, self.view.bounds.size.width-20.0f, 240.0f)];
    
        chartView.backgroundColor = [UIColor clearColor];
    
        NSArray *series = [self.data objectAtIndex:indexPath.row][self.dataName];
        int sum = 0;
        for (NSNumber *number in series) {
            sum += number.integerValue;
        }
        
        options[@"subtitle"] = [NSString stringWithFormat:@"%d %@", sum, options[@"unit"]];
    
        chartView.options = [OptionsProvider provideOptionsForChartType:options series:series type:@"day"];
        
        [cell addSubview:chartView];
    
        [self.charts addObject:chartView];
    
        UIButton *button = [UIButton buttonWithType:UIButtonTypeCustom];
        [button setImage:[UIImage imageNamed:@"ic_info_outline_white"] forState:UIControlStateNormal];
    
        [button setFrame:CGRectMake(self.view.bounds.size.width-20.0f-5.0f-24.0f, 15.0f, 24, 24)];
    
        button.tag = indexPath.row;
        [button addTarget:self action:@selector(showDetailData:) forControlEvents:UIControlEventTouchUpInside];
    
        [chartView addSubview:button];
    }
    else {
        [self updateCellButtonTag:cell newIndex:indexPath.row];
    }
    
    return cell;
}

/*
 // Override to support conditional editing of the table view.
 - (BOOL)tableView:(UITableView *)tableView canEditRowAtIndexPath:(NSIndexPath *)indexPath {
 // Return NO if you do not want the specified item to be editable.
 return YES;
 }
 */

/*
 // Override to support editing the table view.
 - (void)tableView:(UITableView *)tableView commitEditingStyle:(UITableViewCellEditingStyle)editingStyle forRowAtIndexPath:(NSIndexPath *)indexPath {
 if (editingStyle == UITableViewCellEditingStyleDelete) {
 // Delete the row from the data source
 [tableView deleteRowsAtIndexPaths:@[indexPath] withRowAnimation:UITableViewRowAnimationFade];
 } else if (editingStyle == UITableViewCellEditingStyleInsert) {
 // Create a new instance of the appropriate class, insert it into the array, and add a new row to the table view
 }
 }
 */

/*
 // Override to support rearranging the table view.
 - (void)tableView:(UITableView *)tableView moveRowAtIndexPath:(NSIndexPath *)fromIndexPath toIndexPath:(NSIndexPath *)toIndexPath {
 }
 */

/*
 // Override to support conditional rearranging of the table view.
 - (BOOL)tableView:(UITableView *)tableView canMoveRowAtIndexPath:(NSIndexPath *)indexPath {
 // Return NO if you do not want the item to be re-orderable.
 return YES;
 }
 */

#pragma mark - Table view delegate

/*
// In a xib-based application, navigation from a table can be handled in -tableView:didSelectRowAtIndexPath:
- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    // Navigation logic may go here, for example:
    // Create the next view controller.
    
    // Pass the selected object to the new view controller.
    
    // Push the view controller.
}
*/

/*
 #pragma mark - Navigation
 
 // In a storyboard-based application, you will often want to do a little preparation before navigation
 - (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
 // Get the new view controller using [segue destinationViewController].
 // Pass the selected object to the new view controller.
 }
 */

#pragma mark - Additional methods

- (void)dataSourceAdd:(NSDictionary*)dataSource
{
    if (![self.sources containsObject:dataSource]) {
        NSMutableArray *tmp = [NSMutableArray arrayWithArray:self.sources];
        [tmp addObject:dataSource];
        self.sources = [tmp copy];
        
        [[NSUserDefaults standardUserDefaults] setValue:self.sources forKey:@"sources"];
    }
    
    [self loadSourcesAndData];
    [self.tableView reloadData];
}

- (void)dataSourceRem:(NSDictionary*)dataSource
{
    if ([self.sources containsObject:dataSource]) {
        NSMutableArray *tmp = [NSMutableArray arrayWithArray:self.sources];
        [tmp removeObject:dataSource];
        self.sources = [tmp copy];
        
        [[NSUserDefaults standardUserDefaults] setValue:self.sources forKey:@"sources"];
    }
    
    [self loadSourcesAndData];
    [self.tableView reloadData];
}

- (void) updateCellButtonTag: (UITableViewCell*) cell newIndex:(NSInteger) index
{
    for (UIView *view in [cell subviews]) {
        if([view isKindOfClass:[HIChartView class]]) {
            for (UIView *subview in [view subviews]) {
                if([subview isKindOfClass:[UIButton class]]) {
                    subview.tag = index;
                }
            }
        }
    }
}

- (IBAction)actionSegment:(UISegmentedControl*)sender
{
    switch (sender.selectedSegmentIndex) {
        case 0:
            self.dataName = @"day";
            break;
        case 1:
            self.dataName = @"week";
            break;
        case 2:
            self.dataName = @"month";
            break;
        case 3:
            self.dataName = @"year";
            break;
        default:
            break;
    }
    
    [self.tableView reloadData];
}

- (void)loadSourcesAndData
{
    self.data = [NSArray array];
    
    self.sources = [[NSUserDefaults standardUserDefaults] valueForKey:@"sources"];
    
    NSMutableArray *tmpData = [NSMutableArray array];
    for (NSDictionary *source in self.sources) {
        
        NSData *data = [NSData dataWithContentsOfFile:[[NSBundle mainBundle] pathForResource:source[@"source"] ofType:@"json"]];
        NSError *error = nil;
        
        [tmpData addObject:[NSJSONSerialization JSONObjectWithData:data options:kNilOptions error:&error]];
    }
    self.data = [tmpData copy];
}

- (void)showDetailData:(UIButton*)sender
{
    [self.tabBarController setSelectedIndex:1];
    
    DataTableViewController *dataView = [[DataTableViewController alloc] initWithNibName:@"DataTableViewController" bundle:nil];
    
    // Pass the selected object to the new view controller.
    
    NSMutableDictionary *options = [NSMutableDictionary dictionaryWithDictionary:[self.sources objectAtIndex:sender.tag]];
    
    [dataView setConfiguration:options];
    
    UINavigationController *navController = (UINavigationController*)self.tabBarController.selectedViewController;
    if ([navController.visibleViewController isKindOfClass:[DataTableViewController class]]) {
        [navController popViewControllerAnimated:NO];
    }
    
    //Push the view controller
    [navController pushViewController:dataView animated:YES];
}

@end

File: Example/HighFit/HighFit Swift/HighFit/DashboardViewController.swift
//
//  DashboardViewController.swift
//  HighFit
//
//  License: www.highcharts.com/license
//  Copyright © 2019 Highsoft AS. All rights reserved.
//

import UIKit
import Highcharts

class DashboardViewController: UITableViewController, HIChartViewDelegate {
    
    var sources: [[String:Any]]!
    var data: [Any]!
    var dataName: String!
    var charts: [HIChartView]!
    
    private static let sharedInstance = DashboardViewController(nibName: nil, bundle: nil)
    
    public class func sharedDashboard() -> DashboardViewController {
        return sharedInstance
    }
    
    override init(nibName nibNameOrNil: String?, bundle nibBundleOrNil: Bundle?) {
        super.init(nibName: nibNameOrNil, bundle: nibBundleOrNil)
        
        self.title = "Dashboard"
        
        self.tabBarItem.image = UIImage(named: "ic_content_paste_white")
        
        self.loadSourcesAndData()
        
        self.tableView.rowHeight = 240.0
        
        self.tableView.tableHeaderView = UIView(frame: CGRect(x: 0.0, y: 0.0, width: self.tableView.bounds.size.width, height: 0.01))
        
        self.dataName = "day"
        
        self.tableView.tableHeaderView = nil
    }
    
    required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view from its nib.
        
        self.tableView.tableFooterView = UIView(frame: CGRect(x: 0, y: 0, width: self.view.frame.size.width, height: 8.0))
        
        self.tableView.separatorColor = UIColor.clear
        self.tableView.separatorInset = UIEdgeInsets.zero
    }
    
    override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated)
        
        self.tableView.reloadData()
    }
    
    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    
    // MARK: - Table view data source
    
    override func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
        return 250.0
    }
    
    override func numberOfSections(in tableView: UITableView) -> Int {
        return 1
    }
    
    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return self.data.count
    }
    
    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        var options = self.sources[indexPath.row]
        
        let cellReuseIdentifier: String = options["chartType"] as! String
        
        var cell: UITableViewCell? = tableView.dequeueReusableCell(withIdentifier: cellReuseIdentifier)
        
        if cell == nil {
            cell = UITableViewCell(style: .default, reuseIdentifier: cellReuseIdentifier)
            
            cell!.selectionStyle = UITableViewCellSelectionStyle.none
            
            let chartView = HIChartView(frame: CGRect(x: 5.0, y: 5.0, width: self.view.bounds.size.width - 20, height: 240.0))
            chartView.backgroundColor = UIColor.clear
            chartView.delegate = self
            
            let seriesData = self.data[indexPath.row] as! [String: Any]
            let series = seriesData[self.dataName] as! [Int]
            var sum = 0
            for number in series {
                sum += number
            }
            
            options["subtitle"] = "\(sum) \(options["unit"]!)"
            
            chartView.options = OptionsProvider.provideOptions(forChartType: options, series: series, type: "day")
            
            cell!.addSubview(chartView)
            
            self.charts.append(chartView)
            
            let button = UIButton(type: .custom)
            button.setImage(UIImage(named: "ic_info_outline_white"), for: .normal)
            
            button.frame = CGRect(x: self.view.bounds.size.width - 20.0 - 5.0 - 24.0, y: 15.0, width: 24, height: 24)
            
            button.tag = indexPath.row
            button.addTarget(self, action: #selector(self.showDetailData), for: .touchUpInside)
            
            chartView.addSubview(button)
            
            return cell!
        }
        else {
            self.updateCellButtonTag(cell!, newIndex: indexPath.row)
        }
        
        return cell!
    }
    
    /*
     // Override to support conditional editing of the table view.
     override func tableView(_ tableView: UITableView, canEditRowAt indexPath: IndexPath) -> Bool {
     // Return false if you do not want the specified item to be editable.
     return true
     }
     */
    
    /*
     // Override to support editing the table view.
     override func tableView(_ tableView: UITableView, commit editingStyle: UITableViewCellEditingStyle, forRowAt indexPath: IndexPath) {
     if editingStyle == .delete {
     // Delete the row from the data source
     tableView.deleteRows(at: [indexPath], with: .fade)
     } else if editingStyle == .insert {
     // Create a new instance of the appropriate class, insert it into the array, and add a new row to the table view
     }
     }
     */
    
    /*
     // Override to support rearranging the table view.
     override func tableView(_ tableView: UITableView, moveRowAt fromIndexPath: IndexPath, to: IndexPath) {
     
     }
     */
    
    /*
     // Override to support conditional rearranging of the table view.
     override func tableView(_ tableView: UITableView, canMoveRowAt indexPath: IndexPath) -> Bool {
     // Return false if you do not want the item to be re-orderable.
     return true
     }
     */
    
    /*
     // MARK: - Navigation
     
     // In a storyboard-based application, you will often want to do a little preparation before navigation
     override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
     // Get the new view controller using segue.destinationViewController.
     // Pass the selected object to the new view controller.
     }
     */
    
    // MARK: - Additional methods
    
    func dataSourceAdd(_ dataSource: [String: Any]) {
        if !self.sources.contains(where: { $0["source"] as! String == dataSource["source"] as! String } ) {
            var tmp = self.sources!
            tmp.append(dataSource)
            self.sources = tmp
            
            UserDefaults.standard.setValue(self.sources, forKey: "sources")
        }
        
        self.loadSourcesAndData()
        self.tableView.reloadData()
    }
    
    func dataSourceRem(_ dataSource: [String: Any]) {
        if let index = self.sources.index(where: { $0["source"] as! String == dataSource["source"] as! String } ) {
            var tmp = self.sources!
            tmp.remove(at: index)
            self.sources = tmp
            
            UserDefaults.standard.setValue(self.sources, forKey: "sources")
        }
        
        self.loadSourcesAndData()
        self.tableView.reloadData()
    }
    
    private func updateCellButtonTag(_ cell: UITableViewCell, newIndex index: Int) {
        for view in cell.subviews {
            if view is HIChartView {
                for subview in view.subviews {
                    if subview is UIButton {
                        subview.tag = index
                    }
                }
            }
        }
    }
    
    private func loadSourcesAndData() {
        self.data = [Any]()
        
        self.charts = [HIChartView]()
        
        self.sources = UserDefaults.standard.value(forKey: "sources") as! [[String:Any]]
        
        var tmpData = [Any]()
        for source in self.sources {
            
            do {
                if let sourceName = source["source"] as? String,
                    let sourcePath = Bundle.main.path(forResource: sourceName, ofType: "json"),
                    let sourceData = try? Data(contentsOf: URL(fileURLWithPath: sourcePath)),
                    let sourceJson = try JSONSerialization.jsonObject(with: sourceData) as? [String: Any] {
                    tmpData.append(sourceJson)
                }
            } catch {
                print("Error deserializing JSON: \(error)")
            }
        }
        self.data = tmpData
    }
    
    @objc func showDetailData(_ sender: UIButton) {
        self.tabBarController?.selectedIndex = 1
        
        let dataView = DataTableViewController(nibName: "DataTableViewController", bundle: nil)
        
        // Pass the selected object to the new view controller.
        
        let options = self.sources[sender.tag]
        dataView.configuration = options
        
        let navController = self.tabBarController?.selectedViewController as? UINavigationController
        
        if navController?.visibleViewController is DataTableViewController {
            navController?.popViewController(animated: false)
        }
        
        navController?.pushViewController(dataView, animated: true)
    }
    
    // MARK: - HIGChartViewDelegate
    
    func chartViewDidLoad(_ chart: HIChartView!) {
        print("Did load chart \(chart!)")
    }
    
}
